packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -38,27 +90,177 @@ contract FundingRateStore is FundingRateStoreInterface, Testable {
+        _payFinalFees(msg.sender, fundingRateRecord.proposal.finalFee);
+
+        // Send price request
+        _requestOraclePrice(identifier, fundingRateRecord.proposal.time);
+
+        emit DisputedRate(
+            identifier,
+            fundingRateRecord.proposal.rate.rawValue,
+            fundingRateRecord.proposal.time,
+            fundingRateRecord.proposal.proposer,
+            msg.sender
+        );
+
+        // Delete pending proposal and copy into dispute records.
+        fundingRateRecord.proposal.disputer = msg.sender;
+        fundingRateDisputes[identifier][fundingRateRecord.proposal.time] = fundingRateRecord;
+        delete fundingRateRecords[identifier].proposal;
+    }
+
+    /**
+     * @notice Settle a disputed funding rate. The winner of the dispute, either the disputer or the proposer,
+     * will receive a reward plus their final fee bond. This method will also overwrite the current funding rate
+     * with the resolved funding rate returned by the Oracle. Pending funding rates are unaffected by this method.
+     * @dev This will revert if there is no price available for the disputed funding rate.
+     */
+    function settleDispute(bytes32 identifier, uint256 proposalTime) external {
+        FundingRateRecord storage fundingRateDispute = _getFundingRateDispute(identifier, proposalTime);
+
+        // Get the returned funding rate from the oracle. If this has not yet resolved will revert.
+        // If the fundingRateDispute struct has been deleted, then this call will also fail because the proposal
+        // time will be 0.
+        FixedPoint.Signed memory settlementRate = _getOraclePrice(identifier, fundingRateDispute.proposal.time);
+
+        // Dispute was successful if settled rate is different from proposed rate.
+        bool disputeSucceeded = !settlementRate.isEqual(fundingRateDispute.proposal.rate);
+        address proposer = disputeSucceeded
+            ? fundingRateDispute.proposal.disputer
+            : fundingRateDispute.proposal.proposer;
+
+        if (disputeSucceeded) {
+            // If dispute succeeds:
+            // - pay disputer the dispute reward + final fee rebate
+
+            collateralCurrency.safeTransfer(
+                fundingRateDispute.proposal.disputer,
+                fundingRateDispute.proposal.finalFee.rawValue
+            );
+        } else {
+            // If dispute fails:
+            // - pay proposer the proposal reward + final fee rebate
+
+            collateralCurrency.safeTransfer(
+                fundingRateDispute.proposal.proposer,
+                fundingRateDispute.proposal.finalFee.rawValue
+            );
         }
 
-        fundingRateRecord.proposal = Proposal({ rate: rate, time: getCurrentTime(), proposer: msg.sender });
+        // Update current rate to settlement rate.
+        fundingRateRecords[identifier].rate = settlementRate;
+        emit PublishedRate(identifier, settlementRate.rawValue, proposalTime, proposer);
```

Comment:
What if a proposal has come in since this dispute started?

Reply:
It just behaves like a normal proposal. Should we not allow settled disputes to update the funding rate if a normal proposal updates during the dispute?

Reply:
Well wouldn't the proposal that went through while this was being resolved be for a _later_ timestamp, and therefore, be more up to date?

Reply:
Yep I agreed, and this is what we wrote up in the [spec](https://github.com/UMAprotocol/protocol/issues/2061): 

> publishFundingRate: if pending funding rate liveness has expired, then replaces the current funding rate with the pending funding rate, deletes the pending funding rate, and pays the original proposer a reward. If there is no pending funding rate that has expired, but there is a settled disputed funding rate, then sets the result of the dispute as the current funding rate.

I implemented this in this [commit](https://github.com/UMAprotocol/protocol/pull/2147/commits/9d9406265c4932eae0d3b655dfbbcea55df1a987)